x/stakeibc/keeper/transfer.go

```diff
@@ -0,0 +1,214 @@
+	resp, err := k.bankKeeper.AllBalances(ctx, req)
+	if err != nil {
+		return errorsmod.Wrapf(err, "Couldn't look up balances in holding address")
+	}
+
+	var transferErr error;
+	for _, foundCoin := range resp.Balances {
+		if transferErr = k.TransferCoinToReturn(ctx, hostZone, foundCoin); transferErr != nil {
+			k.Logger(ctx).Error(errorsmod.Wrapf(transferErr, "error in token transfer %+v", foundCoin).Error())
+			continue
+		}
+	}
+
+	return transferErr // will be nil if no errors in loop
+}
+
+
+func (k Keeper) TransferCoinToReturn(ctx sdk.Context, hostZone types.HostZone, coin sdk.Coin) error {
+	memo := ""
+	ibcTransferTimeoutNanos := k.GetParam(ctx, types.KeyIBCTransferTimeoutNanos)
+	timeoutTimestamp := uint64(ctx.BlockTime().UnixNano()) + ibcTransferTimeoutNanos
+
+	// build and send an IBC message for each coin to transfer all back to the hostZone
+	msg := ibctypes.NewMsgTransfer(
+		ibctypes.PortID,
+		hostZone.TransferChannelId,
+		coin,
+		hostZone.CommunityPoolHoldingAddress, // from Stride address, unique to each community pool / hostzone
+		hostZone.CommunityPoolReturnIcaAddress, // to ICA controlled address on foreign hub
+		clienttypes.Height{},
+		timeoutTimestamp,
+		memo,
+	)
+
+	msgTransferResponse, err := k.RecordsKeeper.TransferKeeper.Transfer(sdk.WrapSDKContext(ctx), msg)
+	if err != nil {
+		return errorsmod.Wrapf(err, "Error submitting ibc transfer for %+v", coin)
+	} else {
+		//return errorsmod.Wrapf(errors.New(""), "denom is %s", k.GetDenomOnHostZone(coin.Denom, hostZone))
+		result := fmt.Sprintf("Successfully submitted ibctransfer for %+v with response %+v", 
+				coin, msgTransferResponse)
+		k.Logger(ctx).Info(result)
+
+		// If there was no error in sending the transfer msg, store what the transferred denom will be in callback with amount
+		callbackArgs := types.CommunityPoolReturnTransferCallback{
+			HostZoneId: hostZone.ChainId,
+			DenomStride: coin.Denom,
+			IbcDenom: k.GetDenomOnHostZone(coin.Denom, hostZone),
+			Amount: coin.Amount,
+		}
+		callbackArgsBz, err := proto.Marshal(&callbackArgs)
+		if err != nil {
+			return errorsmod.Wrapf(err, "Unable to marshal pool return transfer callback %+v", callbackArgs)
+		}
+
+		// Register a callback by hand when the transfer msg gets the completed ack, different callback for each coin in module
+		k.ICACallbacksKeeper.SetCallbackData(ctx, icacallbackstypes.CallbackData{
+			CallbackKey:  icacallbackstypes.PacketID(msg.SourcePort, msg.SourceChannel, msgTransferResponse.Sequence),
+			PortId:       msg.SourcePort,
+			ChannelId:    msg.SourceChannel,
+			Sequence:     msgTransferResponse.Sequence,
+			CallbackId:   ICACallbackID_CommunityPoolReturn,
+			CallbackArgs: callbackArgsBz,
+		})
+	}
+	return nil
+}
+
+
+// helper function to find the ibc denom on the foreign chain of tokens after transfer from stride  
+func (k Keeper) GetDenomOnHostZone(strideDenom string, hostZone types.HostZone) (ibcDenom string) {
```

Comment:
Since this will only work with tokens that are created on Stride - and the only tokens created on stride our stTokens, I think we could rename this to something like: `GetStTokenIBCDenomOnHost`

Reply:
ah I see you actually have that function below. My rec would be to probably just call the above directly instead

Reply:
This one was written when we were thinking any denom might make it to the holding address, but even with just DYDX and stDYDX, this was trying to unwind the DYDX denom as well -- ie. RedeemStake would output an ibc representation of DYDX in the holding and then this would return "DYDX" because it is going across the channel in the opposite direction to return to its host zone.

Doesn't matter anymore because the 3 epoch plan separates the 2nd epoch transfer to return and the 3rd epoch ICQ->fundcommunitypool.  The ICQ is checking known denoms so this more general function is no longer needed and like you said we only need the ability to figure out the ibc hash for the stDYDX denom on the DYDX chain now.